scipy/special/tests/test_trig.py

```diff
@@ -57,8 +56,123 @@ def test_intermediate_overlow():
+
+    assert_allclose(expected, actual, atol=1e-10)
+
+def test_cospi():
+    x = np.arange(-16, 16, 1 / 16)
+
+    expected = np.cos(np.pi * x)
+    actual = cospi(x)
+
+    assert_allclose(expected, actual, atol=1e-10)
+
+def test_tanpi():
+    x = np.arange(-16, 16, 1 / 16)
+    x = x[x % 0.5 != 0]
+
+    expected = np.tan(np.pi * x)
+    actual = tanpi(x)
+
+    assert_allclose(expected, actual, atol=1e-10)
+
+def test_cotpi():
+    x = np.arange(-16, 16, 1 / 16)
+    x = x[x % 0.5 != 0]
+
+    expected = 1 / np.tan(np.pi * x)
+    actual = cotpi(x)
+
+    assert_allclose(expected, actual, atol=1e-10)
+
+def test_tanpi_complex():
+    x = np.arange(-4, 4, 1 / 16)
+    y = np.array([-32, -16, -2, -1, -0.5, 0.5, 1, 2, 16, 32]) 
+
+    x, y = np.meshgrid(x, y)
+
+    z = x + 1j*y
+
+    expected = np.tan(np.pi * z)
+    actual = tanpi(z)
+
+    assert_allclose(expected.real, actual.real, rtol=1e-7, atol=1e-7)
+    assert_allclose(expected.imag, actual.imag, rtol=1e-7, atol=1e-7)
+
+def test_tanpi_complex_near_pole():
+    eps_list = [
+        1/4, 1/8, 1/16,
+        1/32, 1/64, 1/128, 1/256, 1/512, 1/1024, 1/2048, 
+        1/4096, 1/8192, 1/16384, 1/32768, 1/65536
+    ]
+
+    for eps in eps_list:
+        for x in [-4.5, -3.5, -2.5, -1.5, -0.5, 0.5, 1.5, 2.5, 3.5, 4.5]:
+            z = np.array([
+                complex(eps, eps), complex(eps, 0), complex(eps, -eps),
+                complex(0, eps), complex(0, -eps),
+                complex(-eps, eps), complex(-eps, 0), complex(-eps, -eps)
+            ])
+
+            actual = tanpi(z + x)
+            expected = -1.0 / np.tan(z * np.pi)
+
+            assert_allclose(expected.real, actual.real, rtol=1e-13, atol=1e-13)
+            assert_allclose(expected.imag, actual.imag, rtol=1e-13, atol=1e-13)
+
+def test_cotpi_complex():
+    x = np.arange(-4, 4, 1 / 16)
+    y = np.array([-32, -16, -2, -1, -0.5, 0.5, 1, 2, 16, 32]) 
+
+    x, y = np.meshgrid(x, y)
+
+    z = x + 1j*y
+
+    expected = 1 / np.tan(np.pi * z)
+    actual = cotpi(z)
+
+    assert_allclose(expected.real, actual.real, rtol=1e-7, atol=1e-7)
+    assert_allclose(expected.imag, actual.imag, rtol=1e-7, atol=1e-7)
+
+def test_cotpi_complex_near_pole():
+    eps_list = [
+        1/4, 1/8, 1/16,
+        1/32, 1/64, 1/128, 1/256, 1/512, 1/1024, 1/2048, 
+        1/4096, 1/8192, 1/16384, 1/32768, 1/65536
+    ]
+
+    for eps in eps_list:
+        for x in [-4, -3, -2, -1, 0, 1, 2, 3, 4]:
+            z = np.array([
+                complex(eps, eps), complex(eps, 0), complex(eps, -eps),
+                complex(0, eps), complex(0, -eps),
+                complex(-eps, eps), complex(-eps, 0), complex(-eps, -eps)
+            ])
+
+            actual = cotpi(z + x)
+            expected = 1.0 / np.tan(z * np.pi)
+
+            assert_allclose(expected.real, actual.real, rtol=1e-13, atol=1e-13)
+            assert_allclose(expected.imag, actual.imag, rtol=1e-13, atol=1e-13)
+
+def test_sinpi_zero_sign_ieee754_2008():
```

Comment:
I'd suggest to condense the following cases with scalar input into two cases with array input (`sinpi([-3.0, -2.0, -1.0, -0.0])` and `sinpi([+0.0, +1.0, +2.0, +3.0])`), to save a few lines of code and possibly reveal the sign pattern more clearly. Array comparison can be done with `np.assert_equal`.

Similarly for `cospi`, `tanpi`, `cotpi`.

Reply:
+1 for not having multiple asserts in the same test. If the first fails, the second doesn't even run.